UI: Add WhatsNew Feature for *nix systems

derrod opened the pull request:
### Description

Add support for "What's New" dialogue on macOS and Linux.

ToDo:
- [x] Test on Linux
- [x] Remove temporary test URL
- [x] mbedTLS based implementation for Linux

### Motivation and Context

This feature is used to notify users about changes or important updates, and should be available on other platforms as well.

### How Has This Been Tested?

Tested on a Mac.

### Types of changes

- New feature (non-breaking change which adds functionality) 

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

derrod commented on UI/nix-update/nix-update.cpp:
```diff
@@ -0,0 +1,275 @@
+extern QCef *cef;
+
+#ifndef MAC_WHATSNEW_URL
+#define MAC_WHATSNEW_URL "https://r2.rodney.io/whatsnew.json"
```
The temporary URL https://r2.rodney.io/whatsnew.json is used since it contains the signature in the `X-Signature` header as the real server would.

derrod commented on UI/nix-update/crypto-helpers.hpp:
```diff
@@ -0,0 +1,77 @@
+	0x42, 0x61, 0x35, 0x66, 0x37, 0x4c, 0x6f, 0x4b, 0x38, 0x43, 0x41, 0x77,
+	0x45, 0x41, 0x41, 0x51, 0x3d, 0x3d, 0x0a, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d,
+	0x45, 0x4e, 0x44, 0x20, 0x50, 0x55, 0x42, 0x4c, 0x49, 0x43, 0x20, 0x4b,
+	0x45, 0x59, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0x0a};
```
The certificate here should probably be null-terminated for mbedtls, and it would also make it easier to load on macOS since we can just read it as a C-string.

derrod commented on UI/nix-update/crypto-helpers-cocoa.mm:
```diff
@@ -0,0 +1,58 @@
+
+void get_uuid(std::string &guid)
+{
+	NSString *str = [[NSUUID UUID] UUIDString];
```
Using UUID here instead of just a random string, does that matter for the server @notr1ch?

derrod replied:
Actually I should probably just do what we did for YouTube OAuth and use `QRandom` as a cross-platform way of doing this, possibly with a C++ std fallback if we still want to support Qt versions < 5.10.

derrod commented:
Added a mbedtls based version and CMake adjustments for Linux. The mbedtls path was only checked on macOS so far however.

Now all that remains is some cleanup and adding some error logging to the crypto helpers.

derrod commented:
Verified working on Linux now (at least with Flatpak):
![2022-08-01_18-39-46_nTRYJc](https://user-images.githubusercontent.com/3123295/182200006-db32d9e1-432c-412e-8152-5d38a5d8e089.png)

I believe this is now ready for review, the CMake changes are what I'd like to get a second opinion on the most.

tytan652 commented on CMakeLists.txt:
```diff
@@ -63,6 +63,7 @@ elseif(OS_POSIX)
   if(OS_LINUX)
     option(ENABLE_WAYLAND "Enable building with support for Wayland (Linux)" ON)
     option(BUILD_FOR_PPA "Build for PPA distribution" OFF)
+    option(ENABLE_WHATSNEW_LINUX "Enable WhatsNew dialog on Linux" OFF)
```
If it on by default and not disable-able on Windows and macOS, I think defaulting it to on is fine and even remove the CMake option.

derrod replied:
I kept it off becuase enabling it causes a hard-fail if OBS is compiled without MbedTLS. I could perhaps do something like we do for RTMPS which defaults to an "AUTO" option and then only hard fail if it has been manually set to "ON".

tytan652 replied:
Forgot about this. Let's make it simple, option with default as off as it is is simple.

tytan652 commented on UI/nix-update/nix-update.cpp:
```diff
@@ -0,0 +1,299 @@
+struct QCef;
+extern QCef *cef;
+
+#ifndef MAC_WHATSNEW_URL
+#define MAC_WHATSNEW_URL "https://r2.rodney.io/whatsnew.json"
+#endif
+
+#ifndef LINUX_WHATSNEW_URL
+#define LINUX_WHATSNEW_URL "https://r2.rodney.io/whatsnew.json"
```
Just a `don't forget to change this`.

RytoEX commented on UI/obs-app.cpp:
```diff
@@ -551,7 +551,9 @@ static bool MakeUserDirs()
 		return false;
+#endif
 
+#if defined(_WIN32) || WHATSNEW_ENABLED
```
Can you link me to another place with this syntax? I believe @PatTheMav prefers `defined(WHATSNEW_ENABLED)` over just dropping the raw definition in place.

derrod replied:
This was a mistake, it was just using `#ifdef ..` before and I forgot the second `defined()`

RytoEX reviewed (changes requested):
@PatTheMav is probably better suited to review the CMake stuff, but I took a quick look anyway.

derrod commented on UI/CMakeLists.txt:
```diff
@@ -348,7 +348,7 @@ if(OS_WINDOWS)
   target_compile_definitions(
     obs PRIVATE UNICODE _UNICODE _CRT_SECURE_NO_WARNINGS
-                _CRT_NONSTDC_NO_WARNINGS)
+                _CRT_NONSTDC_NO_WARNINGS WHATSNEW_ENABLED)
```
You cannot disable it. Disabling the Windows updater has been removed from CMake after the infamous OBS 21.1 incident.

derrod commented on UI/CMakeLists.txt:
```diff
@@ -421,9 +423,16 @@ elseif(OS_MACOS)
   target_sources(obs PRIVATE forms/OBSPermissions.ui window-permissions.cpp
                              window-permissions.hpp)
 
+  target_compile_definitions(obs PRIVATE WHATSNEW_ENABLED)
```
Yes, as this code is always compiled. It's only conditional on Linux where mbedtls may not be present.

derrod commented on UI/nix-update/nix-update.cpp:
```diff
@@ -0,0 +1,299 @@
+extern QCef *cef;
+
+#ifndef MAC_WHATSNEW_URL
+#define MAC_WHATSNEW_URL "https://r2.rodney.io/whatsnew.json"
+#endif
+
+#ifndef LINUX_WHATSNEW_URL
+#define LINUX_WHATSNEW_URL "https://r2.rodney.io/whatsnew.json"
```
Changed now. Can be overridden via `LINUX_WHATSNEW_URL` or `MAC_WHATSNEW_URL` for testing. `https://r2.rodney.io/whatsnew.json` provides the last 27.2.4/27.2.3 file and a valid signature for testing, so make sure your OBS build is set to one of these versions.

PatTheMav commented on UI/nix-update/crypto-helpers-mac.mm:
```diff
@@ -0,0 +1,50 @@
+bool VerifySignature(const uint8_t *buf, const size_t len, const uint8_t *sig,
+		     const size_t sigLen)
+{
+	NSString *pubKeyStr =
```
Have you looked at https://developer.apple.com/documentation/security/1395728-secitemimport?language=objc which supports `kSecFormatPEMSequence`?

derrod replied:
I have not, I probably disregarded it while reading documentation because of the "Import" part, but it looks like you can use it to load a key without actually importing it. I'll give it a shot tomorrow.

PatTheMav commented:
That's a whole lot of code just to show a "What's new" page - what's the reason for the dependency on mbedTLS and shipping a public key in code(!) for this?

derrod commented:
> That's a whole lot of code just to show a "What's new" page - what's the reason for the dependency on mbedTLS and shipping a public key in code(!) for this?

This is based on the Windows implementation, which contains the same signature verification: https://github.com/obsproject/obs-studio/blob/master/UI/win-update/win-update.cpp

I think @notr1ch said that this was basically to prevent someone compromising the OBS server from publishing a malicious entry that would get pushed to users.

On Linux it was a matter of either adding a dependency on something like `libkcapi` (Linux Kernel Crypto API wrapper), or using MbedTLS which we already use for RTMPS, so that seemed like the least worst choice.

> We'd also need to differentiate between "news" and actual update changelogs, the latter of which should appear in the Sparkle update notification instead.

"What's New" is only shown after an update to a new version, whereas the changelog in Sparkle or the Windows updater is shown before. It is generally used to highlight specific larger changes or notify users about important updates (e.g. Twitch API deprecation requiring an OBS update). We've also used it to advertise the Survey to users for example.

tytan652 commented:
> what's the reason for the dependency on mbedTLS

On the Linux side there is multiple libs for that (gnutls, OpenSSL) but nothing "by default", we already use mbedTLS for RTMPS.

PatTheMav commented:
FWIW we already ship OBSPublicDSAKey.pem on macOS for Sparkle (_if_ it's enabled in CMake), so we could also ship another PEM as well (if it's not the same key).

derrod commented:
Hardcoding the key does seems somewhat archaic, but I don't know if there's a good way to read a file at compile time (`std::embed` seems to be stuck in limbo, and `#embed` will only come with C23),

That may be a question for @jp9000 as to why this is the way it is and how we should address it.

derrod commented:
Updated now to use `SecItemImport` for Mac, as well as some switches to Qt utilities to replace some old functions.

RytoEX commented:
Looks fine at a glance to me. If @notr1ch and @PatTheMav think the implementation is good enough for now and further improvements/changes can be done later, then we just need to decide when we actually want to merge this.

PatTheMav commented:
> Updated now to use `SecItemImport` for Mac, as well as some switches to Qt utilities to replace some old functions.

I checked "OBSPublicDSAKey.pem" that we ship in the Resources subdirectory, and it's the same key. I'd rather use that key then a hardcoded one. Also the feature requires obs-browser to be present, so configuration should fail if `ENABLE_BROWSER` is disabled for conflicting configuration parameters.

derrod commented:
I don't feel particularly strongly about hardcoding the key either way, so unless there are any objections from R1CH or Jim I can look into loading it from the file. Although that may require additional modifications to CMake/build scripts to include the file in Linux distributions as well.

What's New is - weirdly - not guarded behind the browser being available on Windows, but it's true that it doesn't really make sense for it to be enabled if it isn't, so I'll update that.

derrod commented:
Okay implemented the following changes:

* WhatsNew is now only enabled if browser panels are enabled
  + On Windows/macOS it is implicity enabled
  + On Linux is has to be explicitly enabled as it requires MbedTLS and we probably don't want third-party distributions to have it on by default
* Added updater public key to `UI/data/OBSPublicRSAKey.pem`
* Public key is now loaded from disk (Windows implementation remains unchanged)

> I checked "OBSPublicDSAKey.pem" that we ship in the Resources subdirectory, and it's the same key. 

How did you check? Because they're clearly not the same key. The existing pem file also is a DSA key rather than RSA.

PatTheMav commented on CMakeLists.txt:
```diff
@@ -63,6 +63,7 @@ elseif(OS_POSIX)
   if(OS_LINUX)
     option(ENABLE_WAYLAND "Enable building with support for Wayland (Linux)" ON)
     option(BUILD_FOR_PPA "Build for PPA distribution" OFF)
+    option(ENABLE_WHATSNEW_LINUX "Enable WhatsNew dialog on Linux" OFF)
```
I don't think we need an additional flag for this - make `ENABLE_WHATSNEW` an `INTERNAL` cache variable (set to "ON") on macOS and Windows, make it an `option` on `OS_LINUX`, also defaulting to "ON".

That would make the changes above (explicitly enabling it in the build script and for FlatPak) unnecessary and forces other package maintainers to explicitly turn it off (as desired for Linux).

derrod replied:
Where would be the most appropiate place to set the internal cache variable? `ObsDefaults_{Windows,macOS}` or just at the top of this file?

PatTheMav commented on UI/CMakeLists.txt:
```diff
@@ -363,6 +363,10 @@ if(OS_WINDOWS)
     target_link_options(obs PRIVATE /LARGEADDRESSAWARE)
   endif()
 
+  if(TARGET OBS::browser-panels)
```
Make this global for all OSes and guard behind `TARGET OBS::browser-panels` and `ENABLE_WHATSNEW`.

PatTheMav commented on UI/CMakeLists.txt:
```diff
@@ -424,6 +428,22 @@ elseif(OS_MACOS)
   target_sources(obs PRIVATE forms/OBSPermissions.ui window-permissions.cpp
                              window-permissions.hpp)
 
+  if(TARGET OBS::browser-panels)
```
See above.

PatTheMav commented on UI/CMakeLists.txt:
```diff
@@ -444,6 +464,33 @@ elseif(OS_POSIX)
     target_link_libraries(obs PRIVATE procstat)
   endif()
+
+  if(OS_LINUX AND ENABLE_WHATSNEW_LINUX)
```
Accordingly just check for `ENABLE_WHATSNEW` here.

derrod replied:
I would keep `OS_LINUX` here to avoid issues with FreeBSD at least.

PatTheMav commented on UI/obs-app.cpp:
```diff
@@ -551,7 +551,9 @@ static bool MakeUserDirs()
 		return false;
+#endif
 
+#if defined(_WIN32) || defined(WHATSNEW_ENABLED)
```
With `WHATSNEW_ENABLED` treated as a global define (see above), the specific check for `_WIN32` could be removed.

PatTheMav commented on deps/CMakeLists.txt:
```diff
@@ -6,6 +6,8 @@ if(OS_WINDOWS)
 
   add_subdirectory(blake2)
   add_subdirectory(lzma)
+elseif(OS_MACOS OR ENABLE_WHATSNEW_LINUX)
```
Could be simplified to just check for `ENABLE_WHATSNEW`.

derrod commented:
@tytan652 @GeorgesStavracas Opinions on making this on-by-default for Linux? For RTMPS we use an "auto" setting in case MbedTLS isn't available, so I'm thinking we could mirror this here unless we want package maintainers to have to opt-out explicitly.

GeorgesStavracas commented:
Didn't give a deep review, but from a quick gland, I don't think there should be any build flags controlling this. I think it should always be enabled if a browser source and mbedtls are available.

tytan652 commented:
> Opinions on making this on-by-default for Linux? For RTMPS we use an "auto" setting in case MbedTLS isn't available, so I'm thinking we could mirror this here unless we want package maintainers to have to opt-out explicitly.

I do want package maintainers to have to opt-out explicitly so `ON` by default. RTMPS may need to have a Linux only default change but it's out of scope.

If the maintainer already disabled the browser plugin, this feature should be disabled with it.
Or if the packager don't want to link to mbedTLS at all, a flag is needed to disable the feature.

derrod commented:
Alright, we'll go with enabled by default then. Thanks for your input!

derrod commented:
Updated based on the latest review:
- `ENABLE_WHATSNEW` is now set as an option on Linux/internal cache on macOS/Windows when browser panels are available
- `WHATSNEW_ENABLED` definition is now set when it is enabled
- Linux builds have it enabled by default, will fail if MbedTLS is unavailable
- Fixed some ifdefs that were still using `_WIN32`

PatTheMav commented on UI/CMakeLists.txt:
```diff
@@ -444,6 +471,25 @@ elseif(OS_POSIX)
+      obs_status(
+        FATAL_ERROR
+        "mbedTLS not found, but required for WhatsNew support on Linux")
+    else()
```
Triggering a message with `FATAL_ERROR` acts like a `throw` in CMake - the `else` is not strictly necessary.

derrod replied:
This one must've fallen through the cracks during some refactoring, I removed the else now.

PatTheMav commented:
Finally was able to test this (took a bit of futzing around with the config for it to actually show anything) and got a crash when closing the browser window.

Looks like it requires obs-browser to be updated to include https://github.com/obsproject/obs-browser/pull/374 before this can be merged.

derrod commented:
> Finally was able to test this (took a bit of futzing around with the config for it to actually show anything) and got a crash when closing the browser window.
> 
> Looks like it requires obs-browser to be updated to include [obsproject/obs-browser#374](https://github.com/obsproject/obs-browser/pull/374) before this can be merged.
> 
> EDIT: It's also unfortunate that users are greeted with an empty window at first (CEF showing a blank, white page) instead of waiting for CEF to having finished the page and _then_ opening the Qt window, but I'm afraid that would require rewriting the feature itself.
> 
> FWIW, it takes so long to load for me that I can see some people closing it before actually reading the contents.

Yup since your last version is presumably already set to 27.2.4 and the "WhatsNew" dialog is for 27.2.4 it won't show unless you remove the `LastVersion*` entries from the config first.

Weirdly I never had any crashes on my system, and it also didn't take particularly long to load, but I also already had a browser source active in my scenes, so perhaps my CEF was already "warmed up"?

Looks like the submodule has already been updated to include https://github.com/obsproject/obs-browser/pull/374 in master.

RytoEX commented:
Successfully built and tested on Ubuntu 22.04 running Xorg. What's New dialog pops up successfully. There seems to be a _short_ UI hang for 1-2 seconds.

RytoEX reviewed (approved):
At a glance, seems fine. I could go either way on "WhatsNew dialog" vs. "What's New dialog" in the CMake strings, so I won't nit or block there. Tested earlier on a native build on Ubuntu 22.04 using Xorg, per my comment above. I think this is good enough as-is and we can iterate and make improvements later if needed.
